src/Razor/test/Microsoft.AspNetCore.Razor.LanguageServer.Test/Semantic/SemanticTokenTestBase.cs

```diff
@@ -92,44 +94,66 @@ protected string GetBaselineFileContents(string baselineFileName)
         return semanticFile.ReadAllText();
     }
 
-    private protected async Task<ProvideSemanticTokensResponse> GetCSharpSemanticTokensResponseAsync(
+    private protected async Task<(ProvideSemanticTokensResponse, Dictionary<Range, int[]>)> GetCSharpSemanticTokensResponseAsync(
         string documentText, Range razorRange, bool isRazorFile, int hostDocumentSyncVersion = 0)
     {
         var codeDocument = CreateCodeDocument(documentText, isRazorFile, DefaultTagHelpers);
-        var csharpRange = GetMappedCSharpRange(codeDocument, razorRange);
-        var csharpTokens = Array.Empty<int>();
+        int[][]? csharpTokens = null;
+        var csharpDocumentUri = new Uri("C:\\TestSolution\\TestProject\\TestDocument.cs");
+        var csharpSourceText = codeDocument.GetCSharpSourceText();
 
-        if (csharpRange is not null)
-        {
-            var csharpDocumentUri = new Uri("C:\\TestSolution\\TestProject\\TestDocument.cs");
-            var csharpSourceText = codeDocument.GetCSharpSourceText();
+        await using var csharpServer = await CSharpTestLspServerHelpers.CreateCSharpLspServerAsync(
+            csharpSourceText, csharpDocumentUri, SemanticTokensServerCapabilities, SpanMappingService, DisposalToken);
+
+        var csharpRanges = GetMappedCSharpRanges(codeDocument, razorRange);
 
-            await using var csharpServer = await CSharpTestLspServerHelpers.CreateCSharpLspServerAsync(
-                csharpSourceText, csharpDocumentUri, SemanticTokensServerCapabilities, SpanMappingService, DisposalToken);
+        var perRangeTokens = new Dictionary<Range, int[]>();
+        var responses = new SemanticTokens[csharpRanges.Length];
+        for (var i = 0; i < csharpRanges.Length; i++)
+        {
+            var range = csharpRanges[i];
             var result = await csharpServer.ExecuteRequestAsync<SemanticTokensRangeParams, SemanticTokens>(
                 Methods.TextDocumentSemanticTokensRangeName,
-                CreateVSSemanticTokensRangeParams(csharpRange, csharpDocumentUri),
+                CreateVSSemanticTokensRangeParams(range, csharpDocumentUri),
                 DisposalToken);
 
-            csharpTokens = result?.Data;
+            responses[i] = result;
+            if (result?.Data is not null)
+            {
+                if (!perRangeTokens.ContainsKey(range))
+                {
+                    perRangeTokens[range] = result.Data;
+                }
+            }
         }
 
-        var csharpResponse = new ProvideSemanticTokensResponse(tokens: csharpTokens, hostDocumentSyncVersion);
-        return csharpResponse;
+        csharpTokens = responses.Select(r => r.Data).WithoutNull().ToArray();
```

Comment:
Consider simplifying this whole system, and just having the `perRangeTokens` as returned from here, and put this line that coallesces them all together in the spot where its needed, in the test verification